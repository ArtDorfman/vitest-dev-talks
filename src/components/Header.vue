<template>
  <header>
    <img alt="Vue logo" class="logo" src="@/assets/logo.svg" width="125" height="125" />

    <div class="wrapper">
      <HelloWorld msg="You did it!" />

      <nav>
        <!-- Break snapshot test -->
        <RouterLink id="link" to="/">Home</RouterLink>
        <!-- <RouterLink type="button" to="/about">About</RouterLink> -->
        <RouterLink data-testid="push-to-about" to="/about">About</RouterLink>
      </nav>
    </div>
  </header>
</template>

<script lang="ts" setup>
import { RouterLink } from 'vue-router'
import HelloWorld from '../components/HelloWorld.vue'
</script>
