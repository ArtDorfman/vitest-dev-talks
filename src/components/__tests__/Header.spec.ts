import { describe, it, expect, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import router from '@/router'
import Header from '../Header.vue'

describe('Testing Router links in the Header', () => {
  const mountedHeader = mount(Header, { global: { plugins: [router] } })

  it('it mounts properly', () => {
    expect(mountedHeader).toBeTruthy()
    expect(mountedHeader.text()).toContain('Home')
    expect(mountedHeader.text()).toContain('About')
  })

  it('click the links', () => {
    const push = vi.spyOn(router, 'push')
    mountedHeader.find('a[id="link"]').trigger('click')
    expect(push).toHaveBeenCalledOnce()
    expect(push).toHaveBeenCalledWith('/')

    // mountedHeader.find('a[type="button"]').trigger('click')

    /* Different variants of how to select the link.
    In Cypress best practice is to use data-testid.
    I aggree with that, so here I also add data-testid attribute */
    mountedHeader.find('a[data-testid="push-to-about"]').trigger('click')

    expect(push).toHaveBeenCalledTimes(2)
    expect(push).toHaveBeenCalledWith('/about')
  })

  it('Header snapshot', () => {
    /* We can use snapshots to compare the HTML output of the component with the previous one.  */
    expect(mountedHeader.html()).toMatchSnapshot()
  })
})
