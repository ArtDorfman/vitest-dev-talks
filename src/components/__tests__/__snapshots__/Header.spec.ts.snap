// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing Router links in the Header > Header snapshot 1`] = `
"<header><img alt=\\"Vue logo\\" class=\\"logo\\" src=\\"/src/assets/logo.svg\\" width=\\"125\\" height=\\"125\\">
  <div class=\\"wrapper\\">
    <div data-v-e17ea971=\\"\\" class=\\"greetings\\">
      <h1 data-v-e17ea971=\\"\\" class=\\"green\\">You did it!</h1>
      <h3 data-v-e17ea971=\\"\\"> Youâ€™ve successfully created a project with <a data-v-e17ea971=\\"\\" href=\\"https://vitejs.dev/\\" target=\\"_blank\\" rel=\\"noopener\\">Vite</a> + <a data-v-e17ea971=\\"\\" href=\\"https://vuejs.org/\\" target=\\"_blank\\" rel=\\"noopener\\">Vue 3</a>. What's next? </h3>
    </div>
    <form data-v-e17ea971=\\"\\"><input type=\\"email\\" placeholder=\\"your@email.com\\" required=\\"\\"><button type=\\"submit\\">Submit</button></form>
    <nav>
      <!-- Break snapshot test --><a href=\\"/\\" class=\\"router-link-active router-link-exact-active\\" id=\\"link\\" aria-current=\\"page\\">Home</a><!-- <RouterLink type=\\"button\\" to=\\"/about\\">About</RouterLink> --><a href=\\"/about\\" class=\\"\\" data-testid=\\"push-to-about\\">About</a>
    </nav>
  </div>
</header>"
`;
